### PR TITLE
Mixed Structure Data Loss Bug Resolution

### LINKED ISSUE #200: [BUG] Data loss for repeating together events when instruments are additionally mapped to other event types
# Expected Behavior
All data recorded in the REDCap project should appear in the supertibble

# Current Behavior
When an instrument is mapped to an event that repeats together as well as events that are non-repeating or repeat separately, the data from the repeating together event is not included in the supertibble.

## How to Reproduce the Bug:
See the [Instrument Types REDCap](https://redcap.chop.edu/redcap_v14.4.1/UserRights/index.php?pid=68151)

Screenshot 2024-08-28 at 3 26 45 PM

```r
library(REDCapTidieR)

data <- read_redcap(Sys.getenv("REDCAP_URI"), "77C2C3A96E00246436A9E49974BD5674", allow_mixed_structure = TRUE)

data |>
  extract_tibble("rs_rt") |>
  nrow()

#> [1] 1
# Should be 2

data |>
  extract_tibble("nr_rs_rt") |>
  nrow()

#> [1] 2
# Should be 3
```

### Checklist
Before submitting this issue, please check and verify below that the submission meets the below criteria:

- [x] The issue is atomic
- [x] The issue description is documented
- [x] The issue title describes the problem succinctly
- [ ] Developers are assigned to the issue
- [x] Labels are assigned to the issue

### PR DESCRIPTION
**Pull request by rsh52**
# Description
This PR addresses a bug found where mixed structure databases outputs may result in data loss in instances where a form may be both "repeating separately" and "repeating together".

Side note: Are there any additional things we should add to the [pkgdown site](https://chop-cgtinformatics.github.io/REDCapTidieR/articles/diving_deeper.html#mixed-structure-instruments)?

# Proposed Changes 
List changes below in bullet format:

- Update `convert_mixed_structure()` field under `clean_redcap_long()` to account for missing case
- Update associated mixed structure test to account for case
- Update the [Mixed Structure test REDCap](https://redcap.chop.edu/redcap_v14.4.1/DataEntry/record_status_dashboard.php?pid=65387) with a repeating-together event
- Update `NEWS.md` to be more accurate to present repo

**Screenshots**
If applicable, add screenshots to help explain the proposed changes

### Issue Addressed
**Closes #200**

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [NA] New/revised functions that update downstream outputs have associated static testing files (`.RDS`) updated under `inst/testdata/create_test_data.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR
- [NA] Pre-release package version incremented using `usethis::use_version()` 
   - Using the same version for now under an upcoming 1.2.0 release

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

### CODE REVIEW

File: R/clean_redcap_long.R [hunk 1]
```diff
@@ -399,17 +398,37 @@ distill_repeat_table_long <- function(form_name,
...
-    db_data_long$redcap_repeat_instance <- if_else(update_mask, 1, db_data_long$redcap_repeat_instance)
+    # Create an update mask column to identify which mixed structure rows need updates
+    db_data_long <- db_data_long %>%
+      mutate(
+        update_mask = case_when(
+          # repeat separately instances
+          !is.na({{ field }}) & is.na(.data$redcap_repeat_instance) ~ TRUE,
+          # repeat together instances
+          !is.na({{ field }}) & !is.na(.data$redcap_repeat_instance) & is.na(.data$redcap_repeat_instrument) ~ TRUE,
+          .default = FALSE
```
Comment (ezraporter) [inline R/clean_redcap_long.R:412]: For backwards compatibility. `.default` wasn't introduced until `dplyr 1.1.0`
```suggestion
          TRUE ~ FALSE
```

File: R/clean_redcap_long.R [hunk 2]
```diff
@@ -399,17 +398,37 @@ distill_repeat_table_long <- function(form_name,
...
-    db_data_long$redcap_repeat_instrument <- if_else(update_mask, form, db_data_long$redcap_repeat_instrument)
+    # Assign update data based on rules below
+    db_data_long <- db_data_long %>%
+      mutate(
+        redcap_repeat_instance = case_when(
+          # Add single instance repeat event instance vals when none exist
+          update_mask & is.na(redcap_repeat_instance) ~ 1,
+          # Keep repeat event instance vals when they already exist
+          update_mask & !is.na(redcap_repeat_instance) ~ redcap_repeat_instance,
+          .default = .data$redcap_repeat_instance
```
Comment (ezraporter) [inline R/clean_redcap_long.R:424]:
```suggestion
          TRUE ~ .data$redcap_repeat_instance
```

File: R/clean_redcap_long.R [hunk 3]
```diff
@@ -399,17 +398,37 @@ distill_repeat_table_long <- function(form_name,
...
+        ),
+        redcap_repeat_instrument = case_when(
+          update_mask ~ form,
+          .default = .data$redcap_repeat_instrument
```
Comment (ezraporter) [inline R/clean_redcap_long.R:428]:
```suggestion
          TRUE ~ .data$redcap_repeat_instrument
```

File: R/clean_redcap_long.R [hunk 4]
```diff
@@ -399,17 +398,37 @@ distill_repeat_table_long <- function(form_name,
...
+    form <- mixed_structure_ref$form_name[i] # nolint: object_usage_linter
 
-    # Update redcap_repeat_instance
-    db_data_long$redcap_repeat_instance <- if_else(update_mask, 1, db_data_long$redcap_repeat_instance)
+    # Create an update mask column to identify which mixed structure rows need updates
+    db_data_long <- db_data_long %>%
+      mutate(
+        update_mask = case_when(
+          # repeat separately instances
+          !is.na({{ field }}) & is.na(.data$redcap_repeat_instance) ~ TRUE,
```
Comment (ezraporter) [inline R/clean_redcap_long.R:409]: I don't think `!is.na({{ field }})` is actually working. It always returns `TRUE`. Some NSE and `tidyselect` alternatives:

```r
library(tidyverse)

col <- "x"

tibble(x = c(1, NA)) |>
  mutate(y = if_else(is.na({{col}}), 1, 0))

#># A tibble: 2 × 2
#>      x     y
#>  <dbl> <dbl>
#>1     1     0
#>2    NA     0

# Wrong

tibble(x = c(1, NA)) |>
  mutate(y = if_else(is.na(!!as.symbol(col)), 1, 0))

#># A tibble: 2 × 2
#>      x     y
#>  <dbl> <dbl>
#>1     1     0
#>2    NA     1

# Right

tibble(x = c(1, NA)) |>
  mutate(across(all_of(col), ~if_else(is.na(.), 1, 0), .names = "y"))

#># A tibble: 2 × 2
#>      x     y
#>  <dbl> <dbl>
#>1     1     0
#>2    NA     1

# Right
```

**Review by ezraporter — CHANGES_REQUESTED** [review]
Noted a couple issues in the comments but I think the idea is right!

File: R/clean_redcap_long.R (hunk 1 above)
Comment (rsh52) [reply R/clean_redcap_long.R:412]: Ah yes, good catch. I just hate "TRUE ~ *" that I started to block it from my memory.

File: R/clean_redcap_long.R (hunk 4 above)
Comment (rsh52) [reply R/clean_redcap_long.R:409]: Ah thanks for the great reprex. I'll go for option 2 for now unless you have a strong preference.

**Review by ezraporter — APPROVED** [review]
Looks good now! Thanks for making those changes

**Pull request merged**